File: order_service_consumer/src/main/java/com/github/marceloedudev/packages/sagas/choreography/Step.java
package com.github.marceloedudev.packages.sagas.choreography;

import com.github.marceloedudev.domain.sagas.SagaStep;
import com.github.marceloedudev.packages.queue.MessageHandler;

import java.util.Objects;
import java.util.Optional;

public class Step implements SagaStep {

    private String description;
    private Optional<MessageHandler> invocation;
    private Optional<MessageHandler> compensation;

    public Step(String description) {
        this.description = description;
        this.invocation = Optional.empty();
        this.compensation = Optional.empty();
    }

    @Override
    public void setInvocation(MessageHandler handler) {
        invocation = Optional.of(handler);
    }

    @Override
    public void setCompensation(MessageHandler handler) {
        compensation = Optional.of(handler);
    }

    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public Optional<MessageHandler> getInvocation() {
        return invocation;
    }

    @Override
    public Optional<MessageHandler> getCompensation() {
        return compensation;
    }

    @Override
    public String toString() {
        return "Step{" +
                "description='" + description + '\'' +
                ", invocation=" + invocation +
                ", compensation=" + compensation +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Step step = (Step) o;
        return Objects.equals(description, step.description) && Objects.equals(invocation, step.invocation) && Objects.equals(compensation, step.compensation);
    }

    @Override
    public int hashCode() {
        return Objects.hash(description, invocation, compensation);
    }
}

File: auth_service/src/packages/sagas/choreography/Step.ts
import MessageHandler from "@/packages/eventBus/abstractions/MessageHandler";
import SagaStep from "@/domain/sagas/SagaStep";

export default class Step implements SagaStep {
    #description: string;

    #invocation: MessageHandler | null;

    #compensation: MessageHandler | null;

    constructor(description: string) {
        this.#description = description;
        this.#invocation = null;
        this.#compensation = null;
    }

    setInvocation(handler: MessageHandler) {
        this.#invocation = handler;
    }

    setCompensation(handler: MessageHandler) {
        this.#compensation = handler;
    }

    getDescription(): string {
        return this.#description;
    }

    getInvocation(): MessageHandler | null {
        return this.#invocation;
    }

    getCompensation(): MessageHandler | null {
        return this.#compensation;
    }
}
